scalding-core/src/main/scala/com/twitter/scalding/typed/WritePartitioner.scala

```diff
@@ -215,80 +215,163 @@ object WritePartitioner {
+        case DebugPipe(p) => isLogicalReduce(p)
+        case FilterKeys(p, _) => isLogicalReduce(p)
+        case Filter(p, _) => isLogicalReduce(p)
+        case FlatMapValues(p, _) => isLogicalReduce(p)
+        case FlatMapped(p, _) => isLogicalReduce(p)
+        case ForceToDisk(_) => false // not reducers for sure, could be a map-only job
+        case Fork(_) => false // TODO, not super clear
+        case HashCoGroup(left, _, _) => isLogicalReduce(left)
+        case MapValues(p, _) => isLogicalReduce(p)
+        case Mapped(p, _) => isLogicalReduce(p)
+        case MergedTypedPipe(_, _) => false
+        case ReduceStepPipe(_) => true
+        case SumByLocalKeys(p, _) => isLogicalReduce(p)
+        case TrappedPipe(p, _, _) => isLogicalReduce(p)
+        case CoGroupedPipe(_) => true
+        case WithOnComplete(p, _) => isLogicalReduce(p)
+        case WithDescriptionTypedPipe(p, _) => isLogicalReduce(p)
+      }
+    }
+
+    /**
+     * This is a lattice value that tracks
+     * what we have seen below a given TypedPipe as
+     * we recurse up.
+     */
+    sealed abstract class BelowState {
+      def |(that: BelowState): BelowState =
+        (this, that) match {
+          case (BelowState.Write, later) => later
+          case (BelowState.OnlyMapping, BelowState.Write) => BelowState.OnlyMapping
+          case (BelowState.OnlyMapping, mapOrMater) => mapOrMater
+          case (BelowState.Materialized, _) => BelowState.Materialized
+        }
+    }
+    object BelowState {
+      case object Write extends BelowState
+      case object OnlyMapping extends BelowState
+      case object Materialized extends BelowState
+    }
+    type P[a] = (TypedPipe[a], BelowState)
     // Now we convert
-    val fn = Memoize.functionK[TypedPipe, mat.TP](
-      new Memoize.RecursiveK[TypedPipe, mat.TP] {
+    val fn = Memoize.functionK[P, mat.TP](
+      new Memoize.RecursiveK[P, mat.TP] {
         import TypedPipe._
+        import BelowState._
 
         def toFunction[A] = {
-          case (cp: CounterPipe[a], rec) =>
-            mat.map(rec(cp.pipe))(CounterPipe(_: TypedPipe[(a, Iterable[((String, String), Long)])]))
-          case (c: CrossPipe[a, b], rec) =>
-            rec(c.viaHashJoin)
-          case (cv@CrossValue(_, _), rec) =>
-            rec(cv.viaHashJoin)
-          case (p: DebugPipe[a], rec) =>
-            mat.map(rec(p.input))(DebugPipe(_: TypedPipe[a]))
-          case (p: FilterKeys[a, b], rec) =>
-            mat.map(rec(p.input))(FilterKeys(_: TypedPipe[(a, b)], p.fn))
-          case (p: Filter[a], rec) =>
-            mat.map(rec(p.input))(Filter(_: TypedPipe[a], p.fn))
-          case (Fork(src@IterablePipe(_)), rec) =>
-            // no need to put a checkpoint here:
-            rec(src)
-          case (Fork(src@SourcePipe(_)), rec) =>
+          case ((cp: CounterPipe[a], bs), rec) =>
+            mat.map(rec((cp.pipe, bs | Write)))(CounterPipe(_: TypedPipe[(a, Iterable[((String, String), Long)])]))
```

Comment:
why is a counter pipe `| Write` ?

Reply:
it doesn't need to be, `Write` is the bottom, so `x | Write = x`.